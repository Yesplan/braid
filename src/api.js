import axios from 'axios';

export const getMe = token =>
  axios.get('me', options({ token })).then(response => response.data);

export const getProjects = () => axios.get('projects', options());

export const getStories = id =>
  axios
    .get(`projects/${id}/stories`, options())
    .then(response => response.data);

export const getCurrentIteration = projectId =>
  axios
    .get(`/projects/${projectId}/iterations?scope=current`, options())
    .then(response => response.data);

export const getMemberships = projectId =>
  axios
    .get(`/projects/${projectId}/memberships`, options())
    .then(response => response.data);

export const getBlockers = (projectId, storyIds) =>
  axios
    .get(
      `/projects/${projectId}/stories/bulk`,
      options({ params: { ids: storyIds.join(','), fields: 'blockers' } })
    )
    .then(response => response.data);

const options = ({ token, params } = {}) => ({
  baseURL: 'https://www.pivotaltracker.com/services/v5/',
  headers: { 'X-TrackerToken': token || window.localStorage.getItem('token') },
  params
});
