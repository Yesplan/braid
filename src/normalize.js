import { arrayShuffle, getDayOfYear } from './utils';

const normalizeArray = array => {
  const obj = {};

  array.forEach(element => {
    obj[element.id] = element;
  });

  return obj;
};

const uniqueArray = array =>
  array.filter((value, index, self) => self.indexOf(value) === index);

const removeReleaseStories = story => story.storyType !== 'release';

const normalize = ({ iterationResponse, membershipsResponse }) => {
  const currentIteration = iterationResponse[0];

  const { stories: allStories, ...iteration } = currentIteration;
  const stories = allStories.filter(removeReleaseStories);

  stories.forEach(story => {
    story.blockers = [];
  });

  const people = normalizeArray(membershipsResponse.map(item => item.person));

  const uniqueOwnerIds = arrayShuffle(
    uniqueArray(
      [].concat.apply([], currentIteration.stories.map(story => story.ownerIds))
    ),
    getDayOfYear(new Date())
  );

  return {
    iteration,
    stories,
    people,
    uniqueOwnerIds
  };
};

export default normalize;
