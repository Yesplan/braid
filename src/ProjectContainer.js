import React, { useReducer, useEffect } from 'react';
import { string } from 'prop-types';
import {
  getCurrentIteration,
  getMemberships,
  getBlockers,
  putStory
} from './api';
import { Spinner } from './Spinner';
import { normalize, getStoryIds } from './normalize';
import { Project } from './Project';

const initialState = {
  isFetching: false,
  error: null,
  iteration: {},
  storyIds: [],
  stories: {},
  people: {},
  uniqueOwnerIds: []
};

function reducer(state, action) {
  const { type, payload } = action;

  switch (type) {
    case 'FETCH_ITERATION': {
      return { ...state, isFetching: true };
    }

    case 'FETCH_ITERATION_SUCCESS': {
      return { ...state, isFetching: false, ...normalize(payload) };
    }

    case 'FETCH_BLOCKERS_SUCCESS': {
      const storiesWithBlockers = { ...state.stories };

      payload.forEach(story => {
        if (story.blockers.length > 0) {
          storiesWithBlockers[story.id].blockers = story.blockers;
        }
      });

      return { ...state, stories: storiesWithBlockers };
    }

    case 'FETCH_REQUEST_ERROR': {
      return { ...initialState, error: payload };
    }

    case 'STORY_DROP': {
      return {
        ...state,
        stories: {
          ...state.stories,
          [payload.storyId]: {
            ...state.stories[payload.storyId],
            prevState: state.stories[payload.storyId].currentState,
            currentState: payload.target
          }
        },

        updateStory: {
          ...payload
        }
      };
    }

    case 'STORY_DROP_PERFROM': {
      return {
        ...state,
        updateStory: {
          ...state.updateStory,
          request: true
        }
      };
    }

    case 'STORY_DROP_FAILURE': {
      return {
        ...state,
        stories: {
          ...state.stories,
          [payload.storyId]: {
            ...state.stories[payload.storyId],
            currentState: state.stories[payload.storyId].prevState
          }
        },

        updateStory: null
      };
    }

    case 'STORY_DROP_SUCCESS': {
      return {
        ...state,
        updateStory: null
      };
    }

    default: {
      return { ...state };
    }
  }
}

function ProjectContainer({ id, render }) {
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    const fetchData = async id => {
      try {
        dispatch({ type: 'FETCH_ITERATION' });

        const [iterationResponse, membershipsResponse] = await Promise.all([
          getCurrentIteration(id),
          getMemberships(id)
        ]);

        dispatch({
          type: 'FETCH_ITERATION_SUCCESS',
          payload: { iterationResponse, membershipsResponse }
        });

        const blockers = await getBlockers(id, getStoryIds(iterationResponse));

        dispatch({ type: 'FETCH_BLOCKERS_SUCCESS', payload: blockers });
      } catch (error) {
        dispatch({ type: 'FETCH_REQUEST_ERROR', payload: error });
      }
    };

    fetchData(id);
  }, [id]);

  useEffect(() => {
    const updateStorySideEffect = async ({ storyId, target }) => {
      try {
        dispatch({ type: 'STORY_DROP_PERFROM', payload: { storyId, target } });

        await putStory(id, storyId, { current_state: target });

        dispatch({ type: 'STORY_DROP_SUCCESS', payload: { storyId, target } });
      } catch {
        dispatch({ type: 'STORY_DROP_FAILURE', payload: { storyId, target } });
        alert('Updating story state failed');
      }
    };

    if (state.updateStory && !state.updateStory.request) {
      updateStorySideEffect(state.updateStory);
    }
  });

  if (state.isFetching) {
    return <Spinner />;
  }
  if (state.error) {
    return <div>Error</div>;
  }

  return <Project {...state} dispatch={dispatch} />;
}

ProjectContainer.propTypes = {
  id: string.isRequired
};

export { ProjectContainer };
