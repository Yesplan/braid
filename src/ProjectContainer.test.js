import React from 'react';
import { render, wait } from 'react-testing-library';
import { ProjectContainer } from './ProjectContainer';
import * as api from './api';
import { iterationResponse, membershipsResponse } from './normalize.test';

jest.mock('./utils/getDayOfYear');

const mockBlockers = [
  { id: 563, blockers: [{ resolved: false }] },
  { id: 564, blockers: [{ resolved: true }] }
];

let mockRender;

describe('ProjectContainer', () => {
  beforeEach(() => {
    api.getCurrentIteration = jest.fn(() => Promise.resolve(iterationResponse));
    api.getMemberships = jest.fn(() => Promise.resolve(membershipsResponse));
    api.getBlockers = jest.fn(() => Promise.resolve(mockBlockers));

    mockRender = jest.fn(() => <div>mockRender</div>);
  });

  it('fetches current iteration', async () => {
    render(<ProjectContainer id="1" render={() => <div />} />);

    await wait();
    expect(api.getCurrentIteration).toHaveBeenCalledWith('1');
  });

  it('fetches memberships', async () => {
    render(<ProjectContainer id="1" render={mockRender} />);

    await wait();
    expect(api.getCurrentIteration).toHaveBeenCalledWith('1');
  });

  it('fetches blockers', async () => {
    render(<ProjectContainer id="1" render={mockRender} />);

    await wait();
    expect(api.getBlockers).toHaveBeenCalledWith('1', [563, 564]);
  });

  it('renders fetched and normalized data', async () => {
    const { container } = render(<ProjectContainer id="1" />);

    await wait();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders error on request error', async () => {
    api.getCurrentIteration = jest.fn(() =>
      Promise.reject(new Error('request error'))
    );

    const { container } = render(
      <ProjectContainer id="1" render={mockRender} />
    );

    await wait();
    expect(container.querySelector('div')).toHaveTextContent('Error');
  });
});
