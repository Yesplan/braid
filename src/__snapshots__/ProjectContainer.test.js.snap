// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectContainer renders fetched and normalized data 1`] = `
<section
  class="section"
  style="padding-bottom: 12rem;"
>
  <div
    class="columns"
  >
    <div
      class="column"
      style="min-height: 70vh;"
    >
      <h4
        class="title is-4 has-text-centered"
      >
        Pending
      </h4>
    </div>
    <div
      class="column"
      style="min-height: 70vh;"
    >
      <h4
        class="title is-4 has-text-centered"
      >
        Started
      </h4>
    </div>
    <div
      class="column"
      style="min-height: 70vh;"
    >
      <h4
        class="title is-4 has-text-centered"
      >
        Review
      </h4>
    </div>
    <div
      class="column"
      style="min-height: 70vh;"
    >
      <h4
        class="title is-4 has-text-centered"
      >
        Accepted | Done
      </h4>
      <div
        class="card"
        data-current-state="accepted"
        data-story-id="563"
        data-story-type="feature"
        data-url="http://localhost/story/show/563"
        draggable="true"
      >
        <div
          class="card-content"
        >
          <div
            class="subtitle is-4"
          >
            <a
              class="has-text-grey-dark has-hover-underline"
              href="http://localhost/story/show/563"
              rel="noopener noreferrer"
              target="_blank"
            >
              Complete construction of the Expeditionary Battle Planetoid
            </a>
          </div>
          <div
            class="media"
          >
            <div
              class="media-content"
            >
              <div
                class="tags has-addons is-marginless"
              >
                <div
                  class="tag is-primary"
                  data-testid="feature-tag"
                >
                  <span
                    class="icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-star fa-w-18 "
                      data-icon="star"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 576 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </div>
                <span
                  class="tag is-dark"
                  data-testid="estimate-tag"
                >
                  3
                </span>
                <div
                  class="tag"
                  data-testid="label-tag"
                >
                  plans
                </div>
              </div>
            </div>
            <div
              class="media-right"
            >
              <div
                class="tags is-marginless"
                data-testid="owners"
              >
                <div
                  class="tag is-rounded"
                >
                  <span
                    class="is-uppercase"
                    title="Wilhuff Tarkin"
                  >
                    WT
                  </span>
                </div>
                <div
                  class="tag is-rounded"
                >
                  <span
                    class="is-uppercase"
                    title="Darth Vader"
                  >
                    DV
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="tag is-warning"
          data-testid="blocked-tag"
          style="position: absolute; top: 10px; right: 0px; height: 1.75em; font-size: .65rem; opacity: 1; border-top-right-radius: 0; border-bottom-right-radius: 0;"
        >
          Blocked
        </span>
      </div>
      <div
        class="card"
        data-current-state="accepted"
        data-story-id="564"
        data-story-type="feature"
        data-url="http://localhost/story/show/564"
        draggable="true"
      >
        <div
          class="card-content"
        >
          <div
            class="subtitle is-4"
          >
            <a
              class="has-text-grey-dark has-hover-underline"
              href="http://localhost/story/show/564"
              rel="noopener noreferrer"
              target="_blank"
            >
              Test the Expeditionary Battle Planetoid
            </a>
          </div>
          <div
            class="media"
          >
            <div
              class="media-content"
            >
              <div
                class="tags has-addons is-marginless"
              >
                <div
                  class="tag is-primary"
                  data-testid="feature-tag"
                >
                  <span
                    class="icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-star fa-w-18 "
                      data-icon="star"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 576 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </div>
                <span
                  class="tag is-dark"
                  data-testid="estimate-tag"
                >
                  3
                </span>
                <div
                  class="tag"
                  data-testid="label-tag"
                >
                  plans
                </div>
              </div>
            </div>
            <div
              class="media-right"
            >
              <div
                class="tags is-marginless"
                data-testid="owners"
              >
                <div
                  class="tag is-rounded"
                >
                  <span
                    class="is-uppercase"
                    title="Wilhuff Tarkin"
                  >
                    WT
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="tag is-warning"
          data-testid="blocked-tag"
          style="position: absolute; top: 10px; right: 0px; height: 1.75em; font-size: .65rem; opacity: 0; border-top-right-radius: 0; border-bottom-right-radius: 0;"
        >
          Blocked
        </span>
      </div>
    </div>
  </div>
</section>
`;
